### ISSUE TITLE
iscsiadm errors

### ISSUE THREAD

**Issue by djjudas21** [democratic-csi/democratic-csi#31]
Hi again. I'm struggling with some operational problems with iSCSI. Since I've got democratic-csi and TrueNAS working nicely together with NFS and iSCSI I've been porting my PVCs across. It's all been working great until I hit this one PVC which got provisioned perfectly but for some reason can't be mounted.

I thought it was just a transient error so I deleted that PVC and retried deployment, but the same happened - provisioned successfully but then couldn't mount.

Nothing has changed to my democratic-csi config nor the config on my TrueNAS box.

```
Warning  FailedMount       46s (x7 over 10m)     kubelet            MountVolume.MountDevice failed for volume "pvc-7ed6610c-0f59-4426-a9e9-361c9083636f" : rpc error: code = Internal desc = {"code":19,"stdout":"Logging in to [iface: default, target: iqn.2005-10.org.freenas.ctl:csi-pvc-7ed6610c-0f59-4426-a9e9-361c9083636f-cluster, portal: 192.168.0.4,3260]\n","stderr":"iscsiadm: Could not login to [iface: default, target: iqn.2005-10.org.freenas.ctl:csi-pvc-7ed6610c-0f59-4426-a9e9-361c9083636f-cluster, portal: 192.168.0.4,3260].\niscsiadm: initiator reported error (19 - encountered non-retryable iSCSI login failure)\niscsiadm: Could not log into all portals\n"}
```

I have no idea what I'm doing with `iscsiadm` but on my kubernetes node I did demonstrate that the target in question does show up on the server, but not in the list of sessions:

```
[jonathan@zeus ~]$ sudo iscsiadm -m node
192.168.0.4:3260,4294967295 iqn.2005-10.org.freenas.ctl:csi-pvc-26eba45a-a02f-43a9-ad73-2ef042312ab4-cluster
192.168.0.4:3260,4294967295 iqn.2005-10.org.freenas.ctl:csi-pvc-558510a1-ea93-4454-a40f-3121e190ad11-cluster
192.168.0.4:3260,4294967295 iqn.2005-10.org.freenas.ctl:csi-pvc-676b5237-96c2-4fd1-8522-998fc636f91a-cluster
192.168.0.4:3260,4294967295 iqn.2005-10.org.freenas.ctl:csi-pvc-75e97d92-fa43-4242-bd11-f1e9f1917640-cluster
192.168.0.4:3260,4294967295 iqn.2005-10.org.freenas.ctl:csi-pvc-7ed6610c-0f59-4426-a9e9-361c9083636f-cluster
192.168.0.4:3260,4294967295 iqn.2005-10.org.freenas.ctl:csi-pvc-af8a9ba0-5d2c-41d3-8e5d-72aeb9a15ce5-cluster
192.168.0.4:3260,4294967295 iqn.2005-10.org.freenas.ctl:csi-pvc-ee0f240b-3866-4da9-99b5-c846269aef44-cluster
192.168.0.4:3260,4294967295 iqn.2005-10.org.freenas.ctl:csi-pvc-efb7f43a-93a6-4fff-a13a-d7febcf32c60-cluster
192.168.0.4:3260,4294967295 iqn.2005-10.org.freenas.ctl:csi-pvc-f1f1aad8-a8ce-49dd-af4e-38cbe60a38d1-cluster
192.168.0.4:3260,4294967295 iqn.2005-10.org.freenas.ctl:csi-pvc-fe7ed37f-2ccc-428f-a5f1-b39f8834f7c4-cluster
[jonathan@zeus ~]$ sudo iscsiadm -m session
tcp: [1] 192.168.0.4:3260,1 iqn.2005-10.org.freenas.ctl:csi-pvc-af8a9ba0-5d2c-41d3-8e5d-72aeb9a15ce5-cluster (non-flash)
tcp: [2] 192.168.0.4:3260,1 iqn.2005-10.org.freenas.ctl:csi-pvc-558510a1-ea93-4454-a40f-3121e190ad11-cluster (non-flash)
tcp: [3] 192.168.0.4:3260,1 iqn.2005-10.org.freenas.ctl:csi-pvc-ee0f240b-3866-4da9-99b5-c846269aef44-cluster (non-flash)
tcp: [4] 192.168.0.4:3260,1 iqn.2005-10.org.freenas.ctl:csi-pvc-fe7ed37f-2ccc-428f-a5f1-b39f8834f7c4-cluster (non-flash)
tcp: [6] 192.168.0.4:3260,1 iqn.2005-10.org.freenas.ctl:csi-pvc-676b5237-96c2-4fd1-8522-998fc636f91a-cluster (non-flash)
tcp: [7] 192.168.0.4:3260,1 iqn.2005-10.org.freenas.ctl:csi-pvc-f1f1aad8-a8ce-49dd-af4e-38cbe60a38d1-cluster (non-flash)
```

Have you got any hints for tracking this down? Thanks

**Comment by travisghansen** [comment]
https://github.com/democratic-csi/democratic-csi/issues/12

Smells like that ^ except if you reprovisioned it would have been likely to clean itself up unless when you’re testing a bunch of pvcs are being created simultaneously (ie: deploying a chart with several pvcs).

You can test the theory by running the commands from the `if` block here: https://github.com/democratic-csi/democratic-csi/blob/master/contrib/ctld-config-watchdog-db.sh#L46

**Comment by travisghansen** [comment]
Also how long ago did you deploy the iscsi driver? I’ve made some massive changes over the last week or so that catch some of the undesirable behaviors of the FreeNAS api would could have led to this issue as well. Almost all of them are due to high concurrency operations from creating many pvcs at once though.,

**Comment by travisghansen** [comment]
Also how long ago did you deploy the iscsi driver? I’ve made some massive changes over the last week or so that catch some of the undesirable behaviors of the FreeNAS api would could have led to this issue as well. Almost all of them are due to high concurrency operations from creating many pvcs at once though.,

**Comment by djjudas21** [comment]
Aha yes that sounds like exactly what I'm seeing. In the case I ran into the error, it was with a manifest that created 2 PVCs simultaneously. I didn't notice it creating them individually.

I deployed 0.5.0 with Helm about a week ago. Will update to 0.6.0 if I get some time this evening :+1: 

fwiw, I'm using a mixture of NFS and iSCSI storageclasses, on the same dataset in the same FreeNAS box. Running Postgres over NFS is slloooooooww, so hopefully block storage will be a little more performant. Also considering adding some SSD to TrueNAS

**Comment by travisghansen** [comment]
Ok cool yeah let me know. I reported most of the concurrency bugs upstream to ixsystems and the fixes will likely land in 12.0-U2. I worked around most of them but the bad iscsi config issue is pretty difficult to ‘fix’. I run that script in the background on my server currently to clean up the config in a crude manner at least until the proper fix lands in TrueNAS.

**Comment by travisghansen** [comment]
When you update just make sure you roll the pods to force them to update to latest. I’m pretty sure you’re missing the massive code dumps if you deployed with 0.5.0 though.

I’m very close to snapping a ‘1.0’

**Comment by travisghansen** [comment]
Also, I hope you’re not overlapping directly the nfs and iscsi parent datasets...you don’t want to do that.

**Comment by djjudas21** [comment]
They're on the same dataset but different parents - is this ok?
```
freenas-iscsi.yaml:    datasetParentName: hdd/k8s/iscsi/v
freenas-nfs.yaml:      datasetParentName: hdd/k8s/nfs/vols
```

**Comment by travisghansen** [comment]
Ok yeah that's fine.

**Comment by travisghansen** [comment]
Any luck correcting this behavior?

**Comment by djjudas21** [comment]
I upgraded to 0.6.0 with no problems which seemed to reinstate the ability to create iSCSI PVCs again. I didn't get a chance to extensively test this though because I ran into unrelated problems with my postgres dump file that can't actually be imported due to a problem with constraints, even though it was exported from a healthy database. So long story short I've still got the databases running from NFS where I can just copy in the database files.

Random thought - have you considered writing any components in your project to migrate PVCs between storage classes, especially block storage? Some sort of special job to create/mount the two iSCSI PVs and use something like `dd` to copy the filesystem across?
